thunder/examine/__init__.py

```diff
@@ -308,25 +332,51 @@ def make_trace_dot(trace: TraceCtx):
 
     roots, leaves = bsym_list_to_dag(trace.bound_symbols)
     leaves_id = {id(leaf) for leaf in leaves}
+    roots_id = {id(root) for root in roots}
+
+    def _get_color(node_id):
+        if node_id in roots_id:
+            return "green"
+        if node_id in leaves_id:
+            return "orange"
+        return "lightblue"
+
+    # All roots will be positioned at same level due to `rank`:`same`.
+    roots_g = graphviz.Digraph(graph_attr={"rank": "same"})
+    for root in roots:
+        roots_g.node(str(id(root)), root.bsym.python(indent=0, print_depth=1)[0], fillcolor=_get_color(str(id(root))))
+    dot.subgraph(roots_g)  # Add roots_g to main graph.
+
     stack = [*roots]
     visited = set()
+
+    # Breadth first
     while stack:
-        node: Node = stack.pop()
+        node: Node = stack.pop(0)
         node_id = id(node)
         visited.add(node_id)
-        dot.node(str(node_id), node.bsym.sym.name, fillcolor="orange" if node_id in leaves_id else "white")
-
+        color = _get_color(node_id)
+        dot.node(str(node_id), node.bsym.python(indent=0, print_depth=1)[0], fillcolor=color)
+
```

Comment:
I think here there is the need to do some sanitizing. When trying to make a graph of a model's bwd trace I get `[...]/graphviz/quoting.py:86: expect syntax error scanning invalid quoted string: 'cos, idx, sin, t_lm_head_weight, t_transformer_h_0_attn_attn_weight, \\'`

Reply:
Can you share the snippet where you see this, for a small snippet it seems to work fine. Thanks!

Reply:
here is the diff where I introduced the `make_trace_dot` in the `benchmark_litgpt.py` file

```diff
diff --git a/thunder/benchmarks/benchmark_litgpt.py b/thunder/benchmarks/benchmark_litgpt.py
index 2aa78cbe..8ab1e86d 100644
--- a/thunder/benchmarks/benchmark_litgpt.py
+++ b/thunder/benchmarks/benchmark_litgpt.py
@@ -747,6 +747,11 @@ class Benchmark_litGPT:
             self.perf_metrics["saved_for_backward_tensor_size_mib"] = saved_tensors_size_in_mib
             self.perf_metrics["saved_for_backward_number_of_tensors"] = saved_tensors_len
 
+        from thunder.examine import make_trace_dot
+        bwd_trace = thunder.last_backward_traces(self.model)[-1]
+        dot = make_trace_dot(bwd_trace)
+        dot.render(format='pdf', filename='bwd_trace_graph')
+
     def add_perf_metrics(self):
         if self.throughput:
             metrics = self.throughput.compute()
```

and called the script with:

```shell
python thunder/benchmarks/benchmark_litgpt.py --model_name vicuna-7b-v1.5-16k --compile thunder  --max_iters 3 --warmup_iters 1 --n_layers 2
```

Reply:
Have updated the patch to handle this.